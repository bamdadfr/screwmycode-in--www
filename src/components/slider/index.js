export { SliderComponent } from './slider.component'