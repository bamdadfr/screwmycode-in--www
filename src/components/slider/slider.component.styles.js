import styled, { css } from 'styled-components';

const config = {
  'trackW': '100%',
  // 'trackH': '7px',
  'trackH': '0.5em',
  'thumbD': '1.5em',
  'thumbDV': 1.5,
};

const track = css`
  box-sizing: border-box;
  border: none;
  width: ${config.trackW};
  height: ${config.trackH};
  background: ${(props) => props.theme.background.highlight};
  border-radius: 7px;
  box-shadow: 0 1px 10px 1px ${(props) => props.theme.shadow.opacity};
`;

const progress = css`
  height: ${config.trackH};
  background: ${(props) => props.theme.highlight};
  border-radius: 7px;
`;

const handle = css`
  box-sizing: border-box;
  width: ${config.thumbD};
  height: ${config.thumbD};
  border-radius: 50%;
  border: 3px solid;
  background: ${(props) => props.theme.background.primary};
  box-shadow: 1px 1px 1px ${(props) => props.theme.background.primary},
  0 0 1px ${(props) => props.theme.background.highlight};
`;

const range = (props) => props.max - props.min;
const ratio = (props) => (props.value - props.min) / range (props);

/**
 * @param {object} props react component props
 * @see https://codepen.io/thebabydino/pen/goYYrN
 * @returns {object} styled component
 */
export const Input = styled.input.attrs ((props) => ({
  'style': {
    '--ratio': ratio (props),
  },
}))`
  &, &::-webkit-slider-thumb {
    -webkit-appearance: none;
  }

  margin: 0;
  padding: 0;
  width: ${config.trackW};
  height: ${config.thumbD};
  background: transparent;
  font: 1em/1 arial, sans-serif;

  &:hover {
    cursor: pointer;
  }

  // track
  &::-webkit-slider-runnable-track {
    ${track};
    --sx: calc(.5 * ${config.thumbD} + var(--ratio) * (100% - ${config.thumbD}));
    background: linear-gradient(${(props) => props.theme.highlight}, ${(props) => props.theme.highlight})
      0/ var(--sx) 100% no-repeat ${(props) => props.theme.background.highlight};
  }

  &::-moz-range-track {
    ${track};
  }

  &::-ms-track {
    ${track};
  }

  // progress
  &::-moz-range-progress {
    ${progress};
  }

  &::-ms-fill-lower {
    ${progress};
  }

  // handle
  &::-webkit-slider-thumb {
    margin-top: calc(.5 * (${config.trackH} - ${config.thumbD}));
    ${handle};
  }

  &::-moz-range-thumb {
    ${handle};
  }

  &::-ms-thumb {
    margin-top: 0;
    ${handle};
  }

  &::-ms-tooltip {
    display: none;
  }
`;
