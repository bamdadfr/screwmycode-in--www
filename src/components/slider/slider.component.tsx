import React, {ChangeEvent, ReactElement} from 'react';
import {Input} from './slider.component.styles';

interface SliderComponentProps {
  min: number;
  max: number;
  step: number;
  value: number;
  buffered?: number;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
  name: string;
}

/**
 * Component for a slider
 */
export function SliderComponent({
  min,
  max,
  step,
  value,
  buffered,
  onChange,
  name,
}: SliderComponentProps): ReactElement {
  return (
    <>
      <Input
        type="range"
        title={name}
        aria-label={name}
        min={min}
        max={max}
        step={step}
        value={value}
        buffered={buffered ? buffered : 0}
        onChange={onChange}
      />
    </>
  );
}
