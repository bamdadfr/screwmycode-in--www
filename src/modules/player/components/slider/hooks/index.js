export { useSliderComponent } from './use-slider-component'