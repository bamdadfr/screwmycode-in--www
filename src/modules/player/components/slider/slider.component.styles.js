import styled from 'styled-components'

export const Container = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    margin: 1em;
    background: transparent;
    width: 90%;
`

export const Slider = styled.input`
    width: 100%;

    &::-moz-range-thumb {
        box-shadow: 1px 1px 1px ${(props) => props.theme.background.primary},
        0 0 1px ${(props) => props.theme.background.highlight};
        border: 0.15em solid ${(props) => props.theme.primary};
        background: ${(props) => props.theme.background.highlight};
        cursor: pointer;
        border-radius: 100%;
        width: 1.3em;
        height: 1.3em;
        transition: border 100ms ease-in-out;

        &:hover {
            border: 0.15em solid ${(props) => props.theme.highlight};
        }
    }

    &::-moz-range-track {
        width: 100%;
        height: 8.4px;
        cursor: pointer;
        background: ${(props) => props.theme.background.highlight};
        border-radius: 5px;
        border: 1px solid ${(props) => props.theme.border.primary};
        box-shadow: 0 1px 10px 1px ${(props) => props.theme.shadow.opacity};
        transition: background 100ms ease-in-out;

        &:hover {
            cursor: pointer;
        }
    }
`