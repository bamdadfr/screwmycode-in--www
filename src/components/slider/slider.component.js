import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { StyledSlider, StyledContainer } from './slider.styles'

const propTypes = {
    'value': PropTypes.number.isRequired,
    'handleValue': PropTypes.func.isRequired,
}

/**
 * @function
 * @name SliderComponent
 * @description component: slider
 * @param {object} props - props
 * @param {number} props.value - default value
 * @param {Function} props.handleValue - function from HOC
 * @returns {React.ReactElement} - react component
 */
export function SliderComponent ({ 'value': valueFromProps, handleValue }) {

    const [value, setValue] = useState (valueFromProps.toString ())

    /**
     * @function
     * @name onValue
     * @description pass value to the handler
     */
    function onValue () {

        handleValue (value)

    }

    useEffect (onValue, [value])

    return (
        <>
            <StyledContainer>
                <StyledSlider
                    type="range"
                    aria-label="slider"
                    min={0.5}
                    max={1.5}
                    step={0.005}
                    value={value}
                    onChange={(e) => setValue (e.target.value)}
                />
            </StyledContainer>
        </>
    )

}

SliderComponent.propTypes = propTypes