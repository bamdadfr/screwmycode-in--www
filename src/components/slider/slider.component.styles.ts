import {type SliderComponentProps} from 'src/components/slider/slider.component';
import styled, {css} from 'styled-components';

const config = {
  trackW: '100%',
  // 'trackH': '7px',
  trackH: '0.5em',
  thumbD: '1.5em',
  thumbDV: 1.5,
};

const track = css`
  box-sizing: border-box;
  border: none;
  width: ${config.trackW};
  height: ${config.trackH};
  background: ${(props) => props.theme.background.highlight};
  border-radius: 7px;
  box-shadow: 0 1px 10px 1px ${(props) => props.theme.shadow.opacity};
`;

const progress = css`
  height: ${config.trackH};
  background: ${(props) => props.theme.highlight};
  border-radius: 7px;
`;

const handle = css`
  box-sizing: border-box;
  width: ${config.thumbD};
  height: ${config.thumbD};
  border-radius: 50%;
  border: 3px solid;
  background: ${(props) => props.theme.background.primary};
  box-shadow:
    1px 1px 1px ${(props) => props.theme.background.primary},
    0 0 1px ${(props) => props.theme.background.highlight};
`;

const getProgress = (props: SliderComponentProps) =>
  (props.value - props.min) / (props.max - props.min);
const getBuffered = (props: SliderComponentProps) =>
  (props?.buffered ?? 0 - props.min) / (props.max - props.min);

// @see https://codepen.io/thebabydino/pen/goYYrN
// @ts-expect-error dumb
// noinspection CssUnresolvedCustomProperty
export const Input = styled.input.attrs((props) => ({
  style: {
    '--progress': getProgress(props as SliderComponentProps),
    '--buffered': getBuffered(props as SliderComponentProps),
  },
}))<SliderComponentProps>`
  &,
  &::-webkit-slider-thumb {
    -webkit-appearance: none;
  }

  margin: 0;
  padding: 0;
  width: ${config.trackW};
  height: ${config.thumbD};
  background: transparent;
  font:
    1em/1 arial,
    sans-serif;

  --progress-width: calc(
    0.5 * ${config.thumbD} + var(--progress) * (100% - ${config.thumbD})
  );
  --buffered-width: calc(var(--buffered) * 100%);

  &:hover {
    cursor: pointer;
  }

  // track

  &::-webkit-slider-runnable-track {
    ${track};

    background: linear-gradient(
      to right,
      ${(props) => props.theme.highlight} 0,
      ${(props) => props.theme.highlight} var(--progress-width),
      ${(props) => props.theme.buffered} var(--progress-width),
      ${(props) => props.theme.buffered} var(--buffered-width),
      ${(props) => props.theme.background.highlight} var(--buffered-width),
      ${(props) => props.theme.background.highlight} 100%
    );
  }

  &::-moz-range-track {
    ${track};

    background: linear-gradient(
      to right,
      ${(props) => props.theme.buffered} 0,
      ${(props) => props.theme.buffered} var(--buffered-width),
      ${(props) => props.theme.background.highlight} var(--buffered-width),
      ${(props) => props.theme.background.highlight} 100%
    );
  }

  &::-ms-track {
    ${track};
  }

  // progress

  &::-moz-range-progress {
    ${progress};
  }

  &::-ms-fill-lower {
    ${progress};
  }

  // handle

  &::-webkit-slider-thumb {
    margin-top: calc(0.5 * (${config.trackH} - ${config.thumbD}));
    ${handle};
  }

  &::-moz-range-thumb {
    ${handle};
  }

  &::-ms-thumb {
    margin-top: 0;
    ${handle};
  }

  &::-ms-tooltip {
    display: none;
  }
`;
