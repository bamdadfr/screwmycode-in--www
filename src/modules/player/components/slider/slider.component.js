import React from 'react'
import { Slider, Container } from './slider.component.styles'
import { useSliderComponent } from './hooks'

/**
 * @returns {React.ReactElement} react component
 */
export function SliderComponent () {

    const { value, onChange } = useSliderComponent ()

    return (
        <>
            <Container>
                <Slider
                    type="range"
                    aria-label="slider"
                    min={0.5}
                    max={1.5}
                    step={0.005}
                    value={value}
                    onChange={onChange}
                />
            </Container>
        </>
    )

}