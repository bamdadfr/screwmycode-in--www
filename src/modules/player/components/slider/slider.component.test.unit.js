import React from 'react'
import { SliderComponent } from './slider.component'
import { JestInstance } from '../../../../../jest/jest-instance'

const instance = JestInstance (
    <SliderComponent
        value={1}
        handleValue={jest.fn ()}
    />,
)

it ('should mount and match', () => {

    expect (instance).toBeDefined ()

    expect (instance).toMatchSnapshot ()

})

it ('should have a classname prop from styled components', () => {

    expect (instance.props.className).toBeDefined ()

    expect (instance.props.className).toContain ('sc')

})

it ('should have 1 children', () => {

    expect (instance.children).toHaveLength (1)

})