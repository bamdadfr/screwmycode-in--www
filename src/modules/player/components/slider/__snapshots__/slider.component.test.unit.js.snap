// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should mount and match 1`] = `
<div
  className="sc-1bnwvve-0 einlyy"
>
  <input
    aria-label="slider"
    className="sc-1bnwvve-1 dUGNOc"
    max={1.5}
    min={0.5}
    onChange={[Function]}
    step={0.005}
    type="range"
    value={1}
  />
</div>
`;
